### PR TITLE
add NeptuneLogger [WIP]

### PR DESCRIPTION
**Pull request by kamil-kaczmarek**
## Before submitting (checklist)

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [ ] Did you run [colab minimal CI/CD](https://colab.research.google.com/drive/1JCGTVvWlrIsLXMPRRRSWiAstSLic4nbA) with `latest` and `minimal` requirements?

## Description

* added `NeptuneLogger`
* added `log_artifact` to the logger, so that user can log arbitrary files (not limited to images anymore)

## Type of Change

- [x] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### FAQ
Please review the FAQ before submitting an issue:
- [x] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [x] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [x] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [x] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [x] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

### CODE REVIEW

**Comment by kamil-kaczmarek** [comment]
Opened draft pull request. Note that we are still polishing it :)

File: requirements/requirements-neptune.txt  [hunk 1]
```diff
@@ -0,0 +1 @@
+neptune-client
```
Comment (Scitator) [inline requirements/requirements-neptune.txt ]: could you please also add it to the workflow?
https://github.com/catalyst-team/catalyst/blob/master/.github/workflows/dl_cpu.yml#L98

**Review by Scitator — COMMENTED** [review]
the overall implementation looks great!
could you please resolve the codestyle issues?
```bash
pip install -r requirements/requireemnts-dev.txt
catalyst-make-codestyle
catalyst-check-codestyle
```

File: requirements/requirements-neptune.txt  (hunk 1 above)
Comment (kamil-kaczmarek) [reply requirements/requirements-neptune.txt ]: done.

**Comment by kamil-kaczmarek** [comment]
Hey @Scitator:

1. how can I make `catalyst/tests/test_finetune2.py` to use neptune? Right now we do not test neptune integration in any test. I added this: https://github.com/neptune-ai/catalyst/blob/master/catalyst/tests/test_finetune2.py#L35:L40
Now I need to somehow modify `SETTINGS` so that it will create neptune logger and log this experiment.

2. How can I suppress - `DAR101` in the `core/logger.py` (codestyle checks). I see some errors like this one: https://github.com/catalyst-team/catalyst/pull/1196/checks?check_run_id=2431615565#step:9:111 . They are raised for the `log_artifact`, while they aren't raised for other `log_X` methods.

**Comment by kamil-kaczmarek** [comment]
BTW: please do not merge yet even if all checks pass. We are still working on some proper docs :smiley:

**Comment by Scitator** [comment]
Hi,
1. looks like solved :+1:
2. could you please use a 1-line docstring (there are two lines now)? Yeah, that is a bit strange, but we have a special code style trick for interface classes documentation – 1-line docs shouldn't be reviewed so carefully. If that doesn't help – let's use ``noqa: DAR101``

**Comment by kamil-kaczmarek** [comment]
hey @Scitator ,

ad 1) thanks for the tip -> checks are green now.
ad 2) However, in the `finetune2.py` neptune is not used. We should see catalyst runs in this public project: https://app.neptune.ai/o/common/org/catalyst-integration/experiments?split=tbl&dash=charts&viewId=standard-view - can you help here by any chance?

File: catalyst/loggers/neptune.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        loader_key: str = None,
+        loader_batch_len: int = 0,
+        loader_sample_len: int = 0,
+        loader_batch_step: int = 0,
+        loader_sample_step: int = 0,
+    ) -> None:
+        """Logs batch, epoch and loader metrics to Neptune."""
+        if scope == "batch":
+            neptune_path = "/".join([self.base_namespace, stage_key, loader_key, scope])
+            self._log_metrics(metrics=metrics, neptune_path=neptune_path, step=global_batch_step)
```
Comment (Scitator) [inline catalyst/loggers/neptune.py]:
```suggestion
            self._log_metrics(metrics=metrics, neptune_path=neptune_path, step=global_sample_step)
```

per-sample counter allows you to compare runs with different batch sizes :)

**Comment by Scitator** [comment]
1. 👍 
2. not sure about the error – the overall implementation looks correct. nevertheless, I am not very familiar with anonymized token and how it should work correctly :)

**Comment by kamil-kaczmarek** [comment]
quick note: I'll mention the problem with finetune2 test to our devs.

**Comment by kamil-kaczmarek** [comment]
hey @Scitator, @HubertJaworski from our team will finalize the implemenation today!

File: catalyst/loggers/neptune.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
+from typing import Dict
+
+import numpy as np
+
+from catalyst.core.logger import ILogger
```
Comment (jakubczakon) [inline catalyst/loggers/neptune.py:5]: @HubertJaworski is this correct? ILogger?

**Review by jakubczakon — CHANGES_REQUESTED** [review]

File: catalyst/loggers/neptune.py (hunk 3 above)
Comment (HubertJaworski) [reply catalyst/loggers/neptune.py:5]: @kamil-kaczmarek wrote this and it works correctly

**Review by jakubczakon — APPROVED** [review]

**Comment by HubertJaworski** [comment]
Hi @Scitator 

This PR is ready from our end now.

I'm unable to remove [WIP] tag - can you help?

I managed to fix the test - the loggers were not being added.

I took liberty to add neptune's requirements to pip installations in your CICD and add a check for Neptune's integration in `check_settings.sh`.

I also added some bash functions to improve copy-paste code of pip cleanup & `.catalyst` creation

**Comment by Herudaio** [comment]
@Scitator we were missing changes to CHANGELOG and loggers docs, I've updated that. There are two tests failing, but it seems that it's due to some timeout.

@kamil-kaczmarek created the PR and he is off today (🏝🌴🏝) so I cannot update the checklist :)

File: setup.py [hunk 4]
```diff
@@ -95,9 +95,11 @@ def run(self):
     "onnx": load_requirements("requirements/requirements-onnx.txt"),
     "onnx-gpu": load_requirements("requirements/requirements-onnx-gpu.txt"),
     "mlflow": load_requirements("requirements/requirements-mlflow.txt"),
+    "neptune": load_requirements("requirements/requirements-neptune.txt"),
 }
-extras["all"] = extras["cv"] + extras["ml"] + extras["hydra"] + extras["optuna"]
-
+extras["all"] = (
+    extras["cv"] + extras["ml"] + extras["hydra"] + extras["optuna"] + extras["neptune"]
+)
```
Comment (Scitator) [inline setup.py]:
```suggestion
extras["all"] = extras["cv"] + extras["ml"] + extras["hydra"] + extras["optuna"]
```

**Pull request merged**